### PR TITLE
Add logging for functions in main.py

### LINKED ISSUE #10: Use logger for reporting error or debugging mode
Use logger for reporting error or debugging mode.

In order to:
- improve error handling (eg parallel execution)
- used within library functions
- avoid cluttering user's stdout/stderr

### LINKED ISSUE #354: Save log of terminal printout during training
It would be useful to save a log file of the info printed out in the terminal during training. This file could go into the `"log_directory": ` field specified in the config.json file. Intuitively, I would expect to see such log file, given the name of the json field.

### PR DESCRIPTION
**Pull request by cakester**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [ ] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

Add logging of stdout in main function, so we would have the output print to the console and also logged to file. Note 
 that for the `log_directory` tag, we cannot use it since it is one of the `key_change_list` property under config, meaning that if you use that key it will get overwritten, so we have to use `log_file` key instead. This has been tested by running the main function on itself.

Also, I just updated pr so that it removes all of the prints from ivadomed folder, and uses logger.info instead, also tqdm calls were tried to logged as well

## Linked issues

resolves https://github.com/ivadomed/ivadomed/issues/354
also resolves https://github.com/ivadomed/ivadomed/issues/10

### CODE REVIEW

File: ivadomed/inference.py [hunk 1]
```diff
@@ -194,7 +195,7 @@ def segment_volume(folder_model, fname_images, gpu_id=0, options=None):
 
     # Force filter_empty_mask to False if fname_roi = None
     if fname_roi is None and 'filter_empty_mask' in loader_params["slice_filter_params"]:
-        print("\nWARNING: fname_roi has not been specified, then the entire volume is processed.")
+        logger.info("\nWARNING: fname_roi has not been specified, then the entire volume is processed.")
```
Comment (charleygros) [inline ivadomed/inference.py:198]: Would it be relevant to use `logger.warning` in such places?

File: ivadomed/loader/adaptative.py [hunk 2]
```diff
@@ -89,9 +90,9 @@ def save(self, path):
         """
         try:
             self.df.to_csv(path, index=False)
-            print("Dataframe has been saved at {}.".format(path))
+            logger.info("Dataframe has been saved at {}.".format(path))
         except FileNotFoundError:
-            print("Wrong path.")
+            logger.info("Wrong path.")
```
Comment (charleygros) [inline ivadomed/loader/adaptative.py:95]: Same question as above with `logger.error`

File: ivadomed/config/config_default.json [hunk 3]
```diff
@@ -4,6 +4,7 @@
     "path_output": "path_output",
     "model_name": "my_model",
     "debugging": false,
+    "log_file": "log",
```
Comment (charleygros) [inline ivadomed/config/config_default.json:7]: Could you please add documentation about this parameter? Cheers

File: ivadomed/main.py [hunk 4]
```diff
@@ -308,6 +308,10 @@ def run_command(context, n_gif=0, thr_increment=None, resume_training=False):
     """
     command = copy.deepcopy(context["command"])
     path_output = set_output_path(context)
+    log_file = context['log_file']
```
Comment (charleygros) [inline ivadomed/main.py:311]: Sorry, I do not understand the difference of this line vs [this one](https://github.com/ivadomed/ivadomed/blob/34551756dd6cc3cbec7800a691683096d6a7305b/ivadomed/main.py#L480).

Comment (charleygros) [reply ivadomed/main.py:311]: Also, would be good to have the log file saved within the log_directory, is it the case?

**Review by charleygros — CHANGES_REQUESTED** [review]

File: ivadomed/main.py (hunk 4 above)
Comment (cakester) [reply ivadomed/main.py:311]: no, but I'll change it now

Comment (cakester) [reply ivadomed/main.py:311]: > Sorry, I do not understand the difference of this line vs [this one](https://github.com/ivadomed/ivadomed/blob/34551756dd6cc3cbec7800a691683096d6a7305b/ivadomed/main.py#L480).

the one you mentioned is version_info.log, this one is the log for all logger.info/warning/etc

File: ivadomed/inference.py (hunk 1 above)
Comment (cakester) [reply ivadomed/inference.py:198]: > Would it be relevant to use `logger.warning` in such places?

yes you are right

File: ivadomed/loader/adaptative.py (hunk 2 above)
Comment (cakester) [reply ivadomed/loader/adaptative.py:95]: > Same question as above with `logger.error`

will change

**Comment by cakester** [comment]
right now all logs are going to be saved within log_directory (which is the path_output), the log_file key inside the config will indicate the name of the file the log gets stored to, documented in docs/source/configuration_file.rst

**Comment by jcohenadad** [comment]
> documented in docs/source/configuration_file.rst

@cakester can you pls point to the exact line where it is documented? (quicker than having to search for it)-- thx

File: ivadomed/main.py [hunk 5]
```diff
@@ -308,6 +308,10 @@ def run_command(context, n_gif=0, thr_increment=None, resume_training=False):
     """
     command = copy.deepcopy(context["command"])
     path_output = set_output_path(context)
+    log_file = os.path.join(context['path_output'], context['log_file'])
+    logger.remove()
```
Comment (cakester) [inline ivadomed/main.py:312]: note this is to prevent multiple threads all registering the same sink multiple times (double/tripple logging)

**Comment by cakester** [comment]
> > documented in docs/source/configuration_file.rst
> 
> @cakester can you pls point to the exact line where it is documented? (quicker than having to search for it)-- thx

https://github.com/ivadomed/ivadomed/pull/764/files#diff-b08ae2b1ad620b87870e3f95b73aaa50912520096e71b13356809a822147c98cR138

File: ivadomed/inference.py [hunk 6]
```diff
@@ -173,7 +174,7 @@ def segment_volume(folder_model, fname_images, gpu_id=0, options=None):
         # TRANSFORMATIONS
         # If ROI is not provided then force center cropping
         if fname_roi is None and 'ROICrop' in context["transformation"].keys():
-            print(
+            logger.info(
```
Comment (dyt811) [inline ivadomed/inference.py:177]: This should be a logger.warning

File: ivadomed/loader/adaptative.py [hunk 7]
```diff
@@ -566,17 +567,17 @@ def load_into_ram(self, contrast_lst=None):
         with h5py.File(self.path_hdf5, "r") as hdf5_file:
             for ct in contrast_lst:
                 if ct not in keys:
-                    print("Key error: status has no key {}".format(ct))
+                    logger.info("Key error: status has no key {}".format(ct))
```
Comment (dyt811) [inline ivadomed/loader/adaptative.py:570]: logger.error here

**Review by dyt811 — CHANGES_REQUESTED** [review]
All looks good. Please correct the few logging levels. 

On a related note, could you determine IF we need to adjust the logger level/sink file location, etc, do we have to use the same logger instance object or can configure it at loguru level? This will determine whether we need to use a centralized logger object or not.

File: ivadomed/inference.py [hunk 8]
```diff
@@ -194,7 +195,7 @@ def segment_volume(folder_model, fname_images, gpu_id=0, options=None):
 
     # Force filter_empty_mask to False if fname_roi = None
     if fname_roi is None and 'filter_empty_mask' in loader_params["slice_filter_params"]:
-        print("\nWARNING: fname_roi has not been specified, then the entire volume is processed.")
+        logger.warning("\nWARNING: fname_roi has not been specified, then the entire volume is processed.")
```
Comment (dyt811) [inline ivadomed/inference.py:198]: Can remove "\nWARNING: " here. Please update this in other places.

File: ivadomed/loader/adaptative.py [hunk 9]
```diff
@@ -77,9 +78,9 @@ def load_dataframe(self, path):
         """
         try:
             self.df = pd.read_csv(path)
-            print("Dataframe has been correctly loaded from {}.".format(path))
+            logger.info("Dataframe has been correctly loaded from {}.".format(path))
         except FileNotFoundError:
-            print("No csv file found")
+            logger.info("No csv file found")
```
Comment (dyt811) [inline ivadomed/loader/adaptative.py:83]: Please update this to error.

File: ivadomed/loader/adaptative.py [hunk 10]
```diff
@@ -430,12 +431,12 @@ def _load_filenames(self):
...
-                print(len(input_volumes))
-                print("grp= ", str(subject_id))
+                logger.info(len(input_volumes))
+                logger.info("grp= ", str(subject_id))
                 key = "inputs/{}".format(contrast)
-                print("key = ", key)
+                logger.info("key = ", key)
                 if len(input_volumes) < 1:
-                    print("list empty")
+                    logger.info("list empty")
```
Comment (dyt811) [inline ivadomed/loader/adaptative.py:439]: Can you update to warning here?

**Review by dyt811 — CHANGES_REQUESTED** [review]

File: ivadomed/training.py [hunk 11]
```diff
@@ -274,10 +274,13 @@ def train(model_params, dataset_train, dataset_val, training_params, path_output
...
             if training_params["loss"]["name"] != "DiceLoss":
                 msg += "\tDice validation loss: {:.4f}.".format(val_dice_loss_total_avg)
+            logger.info(msg)
             tqdm.write(msg)
             end_time = time.time()
             total_time = end_time - start_time
-            tqdm.write("Epoch {} took {:.2f} seconds.".format(epoch, total_time))
+            msg = "Epoch {} took {:.2f} seconds.".format(epoch, total_time)
+            logger.info(msg)
+            tqdm.write(msg)
```
[the remark this answers is not in the document]
Comment (cakester) [reply ivadomed/training.py:283]: it doesn't have to I'll change it

**Review by dyt811 — APPROVED** [review]

**Review by charleygros — APPROVED** [review]

**Pull request merged**